test/jdk/java/net/httpclient/lib/jdk/httpclient/test/lib/http2/Http2TestServerConnection.java

```diff
@@ -729,6 +729,13 @@ headers, rspheadersBuilder, uri, bis, getSSLSession(),
                     }
                 }
                 throw closed;
+            } finally {
+                // If the handler reads the exchange's InputStream, this code will have no effect. If the handler does not
+                // read the exchange's InputStream and the send window has been exceeded, then a RST_STREAM frame with
+                // the error code NO_ERROR will be sent to the client before closing.
+                if (bis instanceof BodyInputStream inputStream && (!inputStream.isEof() || inputStream.q.size() > 0)) {
+                    bos.sendResetOnClose(ResetFrame.NO_ERROR);
```

Comment:
I share @dfuch concern. This line will have no effect if `bos` is already closed at this point. We still need to reset the stream to let the peer know that we won't consume `bis`.

Reply:
Thanks for the feedback Daniel. I see the issue that my implementation causes if the handler closes the `bos` early. Working on this now as a priority, solution will try to make sure that this RST_STREAM is sent even if the handler looks like the example you gave above. Probably worth inluding these handlers as test cases also.

Reply:
So to address this in the most recent commit, the BodyOutputStream initalised in the try-with-resources block is now passed a reference to the given BodyInputStream and it is used to check if a reset is required there once it is closed. This is done with the following in BodyOutputStream.java:

```
private boolean resetNeeded() throws IOException {
        return (bis != null && (!bis.isEof() || bis.q.size() > 0));
}
```

For the `sendResponseHeaders(200, -1)` case, a reset has to be triggered to send instead once the BodyOutputStream is closed at the beginning of the exchange (because of the negative content length).